dipy/denoise/denspeed.pyx

```diff
@@ -153,49 +157,70 @@ cdef double process_block(double [:, :, ::1] arr,
 
     cdef:
         cnp.npy_intp m, n, o, M, N, O, a, b, c, cnt, step
-        double patch_vol_size
+        double patch_vol_size, block_vol_size
         double summ, d, w, sumw, sum_out, x, sigm
         double * W
         double * cache
         double * sigma_block
         double denom
         cnp.npy_intp BS = B * 2 + 1
+        cnp.npy_intp PS = P * 2 + 1
+        cnp.npy_intp TS = PS + BS - 1
 
     cnt = 0
     sumw = 0
     patch_vol_size = (P + P + 1) * (P + P + 1) * (P + P + 1)
+    block_vol_size = (B + B + 1) * (B + B + 1) * (B + B + 1)
 
-    W = <double *> malloc(BS * BS * BS * sizeof(double))
-    cache = <double *> malloc(BS * BS * BS * sizeof(double))
-    sigma_block = <double *> malloc(BS * BS * BS * sizeof(double))
+    W = <double *> malloc(PS * PS * PS * sizeof(double))
```

Comment:
But what you want are differences between patches of size PS for all **patches** within BS. There should be no comparing over distances between **blocks**.  This is what the code was doing before.  I think the bug in the old code was just that the block radius searched over was smaller than specified by one patch radius, not in the actual implementation.